### DISCUSSION TITLE
Using SQL to join 3 tables

### DISCUSSION THREAD

**Discussion by ericbaek** [pocketbase/pocketbase#4591] [Q&A] [answered]
let's say I am creating following structured backend system for ticketing website.
![20240319 - New Discussion · pocketbasepocketbase (Arc)@2x](https://github.com/pocketbase/pocketbase/assets/43541203/c583f162-17e6-4310-9d40-410387cfbd05)

I would like to make a custom api route to fetch all information by joining User, Ticket, and Event table.
```
routerAdd("GET", "/user/all/tickets", (c) => {

    const result = arrayOf(new DynamicModel({
        "User.name": "",
        "Ticket.quantity": 0,
        "Event.eventName": "",
    }))
    
    $app.dao().db()
        .select("User.*, Ticket.*, Event.*")
        .from("User")
        .innerJoin("Ticket", $dbx.exp("Ticket.User_rel = User.id"))
        .innerJoin("Event", $dbx.exp("Ticket.Event_rel = Ticket.id"))
        .all(result)

    return c.json(200, result)
})
```

Based on my understanding of the document, I have created the above code. However, it appears to be incorrect. I have several questions regarding this matter.

Firstly, what would be the appropriate code for this scenario? Secondly, should I must include the entire initial data of the DynamicModel? There are approximately 20 fields that need to be filled in. And if so, what will be the correct structure to create for expanded(Joined) data?

**Comment by ganigeorgiev** [comment, marked as answer]
If you just want to pick the 3 columns shown in your code sample, then you can try with something like:

```js
const result = arrayOf(new DynamicModel({
    "name":      "",
    "quantity":  0,
    "eventName": "",
}))

$app.dao().db()
    .select("User.name, Ticket.quantity, Event.eventName") // if it doesn't work try to specify exact aliases to the DynamicModel columns
    ...
```

An alternative to one big query could be to split it in several smaller ones and combine the results programmatically. This approach is usually useful when you have to do additional transformation or grouping on the retrieved data. [N+1 queries are not a problem in SQLite](https://sqlite.org/np1queryprob.html) (_at least for the majority of the cases and sometimes could be even more efficient as it will avoid large cartesian products when performing joins with large datasets_). For example:

```js
const result = [];

const tickets = arrayOf(new Record)
$app.dao().recordQuery("Ticket").all(tickets)

for (const ticket of tickets) {
    // find the ticket user and event
    const user  = $app.dao().findRecordById("User", ticket.get("User_rel"))
    const event = $app.dao().findRecordById("Event", ticket.get("Event_rel"))

    result.push({
        name:      user.get("name"),
        quantity:  ticket.get("quantity"),
        eventName: event.get("eventName"),
    })
}

console.log(result)
```
